toxcore/group_announce.c

```diff
@@ -0,0 +1,458 @@
+/* SPDX-License-Identifier: GPL-3.0-or-later
+ * Copyright © 2016-2020 The TokTok team.
+ * Copyright © 2015 Tox project.
+ */
+#include "group_announce.h"
+#include "LAN_discovery.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "mono_time.h"
+#include "util.h"
+
+/* Minimum size of an announce. */
+#define GCA_ANNOUNCE_MIN_SIZE (ENC_PUBLIC_KEY_SIZE + 2)
+
+/**
+ * Removes `announces` from `gc_announces_list`.
+ */
+static void remove_announces(GC_Announces_List *gc_announces_list, GC_Announces *announces)
+{
+    if (announces == nullptr || gc_announces_list == nullptr) {
+        return;
+    }
+
+    if (announces->prev_announce) {
+        announces->prev_announce->next_announce = announces->next_announce;
+    } else {
+        gc_announces_list->root_announces = announces->next_announce;
+    }
+
+    if (announces->next_announce) {
+        announces->next_announce->prev_announce = announces->prev_announce;
+    }
+
+    free(announces);
+}
+
+/**
+ * Returns the announce designated by `chat_id`.
+ * Returns null if no announce is found.
+ */
+static GC_Announces *get_announces_by_chat_id(const GC_Announces_List *gc_announces_list,  const uint8_t *chat_id)
+{
+    GC_Announces *announces = gc_announces_list->root_announces;
+
+    while (announces) {
+        if (memcmp(announces->chat_id, chat_id, CHAT_ID_SIZE) == 0) {
+            return announces;
+        }
+
+        announces = announces->next_announce;
+    }
+
+    return nullptr;
+}
+
+/**
+ * Adds a maximum of `max_nodes` announces to `gc_announces` for the group designated by `chat_id`.
```

Comment:
This comment line doesn't quite match the one in the header. Why do we (almost identically) duplicate comments in .h and .c? They get out of sync, as we can see here.